# Phase 1: Data Exploration & Initial Assessment

#   Importing libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Load Dataset

In [2]:
df=pd.read_csv('messy_retail_dataset_120k.csv')
df

customer_id   age  gender       city  purchase_amount payment_mode  \
0              2145  33.0  Female       pune          1625.84   debit card   
1             41963  44.0     NaN    Chennai          5239.85  CREDIT CARD   
2             12788  62.0    Male    Chennai              NaN       Debit    
3             31429   NaN       m        NaN          2567.63   NetBanking   
4              2460  25.0  FEMALE        NaN          7671.78          NaN   
...             ...   ...     ...        ...              ...          ...   
124995        48753  38.0     NaN       pune          2070.95  Credit Card   
124996        26416  56.0       m      delhi          7099.90  credit card   
124997        25969  57.0       f       PUNE          5096.26  credit card   
124998        32728  20.0    male  Bangalore            96.07   debit card   
124999        37114  51.0       m   Banglore           556.31         cash   

                        purchase_date  
0       2022-06-26 23:59:05.269671100  
1       2022-12-24 07:17:27.411198698  
2       2022-03-22 13:38:27.803018318  
3       2022-07-21 09:01:18.680025577  
4       2021-03-28 19:47:54.746926192  
...                               ...  
124995  2021-05-17 22:31:31.834594690  
124996  2023-07-25 19:25:18.401680681  
124997  2023-05-28 11:43:10.337243115  
124998  2021-01-02 22:10:58.895947510  
124999  2022-12-02 12:33:28.267668670  

[125000 rows x 7 columns]

# Observation
The dataset contains 125,000 customer purchase records, each representing a 
single transaction with 7 attributes such as age, gender, city, payment mode, 
and purchase amount.


In [3]:
df.head()

customer_id   age  gender     city  purchase_amount payment_mode  \
0         2145  33.0  Female     pune          1625.84   debit card   
1        41963  44.0     NaN  Chennai          5239.85  CREDIT CARD   
2        12788  62.0    Male  Chennai              NaN       Debit    
3        31429   NaN       m      NaN          2567.63   NetBanking   
4         2460  25.0  FEMALE      NaN          7671.78          NaN   

                   purchase_date  
0  2022-06-26 23:59:05.269671100  
1  2022-12-24 07:17:27.411198698  
2  2022-03-22 13:38:27.803018318  
3  2022-07-21 09:01:18.680025577  
4  2021-03-28 19:47:54.746926192

#  Step 1: Data Understanding

# Columns

In [4]:
df.columns

Index(['customer_id', 'age', 'gender', 'city', 'purchase_amount',
       'payment_mode', 'purchase_date'],
      dtype='object')

# Observation
The dataset consists of a mix of numerical and categorical features. 
The numerical columns include customer_id, age, and purchase_amount, 
which provide quantitative information about each transaction. 
The categorical columns—gender, city, and payment_mode—describe and purchase_data customer attributes and purchase behavior

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125000 entries, 0 to 124999
Data columns (total 7 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   customer_id      125000 non-null  int64  
 1   age              118220 non-null  float64
 2   gender           111186 non-null  object 
 3   city             113665 non-null  object 
 4   purchase_amount  112406 non-null  float64
 5   payment_mode     113597 non-null  object 
 6   purchase_date    125000 non-null  object 
dtypes: float64(2), int64(1), object(4)
memory usage: 6.7+ MB


#  Observation
Overall structure of dataset

# Statistical Summary of Numerical Columns

In [6]:
df.describe()

customer_id            age  purchase_amount
count  125000.000000  118220.000000    112406.000000
mean    25433.144056      43.552259      5022.071431
std     14161.932940      15.024888      2868.669526
min      1000.000000      18.000000        50.060000
25%     13144.000000      31.000000      2533.187500
50%     25411.000000      44.000000      5033.690000
75%     37713.000000      57.000000      7504.235000
max     49999.000000      69.000000      9999.920000

# Observtion
The numerical summary shows that customer_id, age, and purchase_amount have strong variability, 
with age averaging around 44 years and purchase amounts averaging about ₹5,022. Although the ranges 
are wide, the values appear reasonable, but missing entries in age and purchase_amount indicate the need for further cleaning.

# Finding Null Values

In [7]:
df.isnull().sum()

customer_id            0
age                 6780
gender             13814
city               11335
purchase_amount    12594
payment_mode       11403
purchase_date          0
dtype: int64

# Observation
Several columns contain missing data, with gender, city, payment_mode, and purchase_amount 
having over 11,000 missing entries each, and age missing in 6,780 records. 
customer_id and purchase_date are fully complete with no missing values

# Understand categories


In [8]:
df['gender'].value_counts()

gender
Female    14088
FEMALE    14038
M         13946
f         13847
Male      13837
male      13832
m         13814
F         13784
Name: count, dtype: int64

In [9]:
df['city'].value_counts()

city
 delhi       11518
PUNE         11456
HYDERABAD    11426
 Banglore    11423
Mumbai       11366
Bangalore    11352
pune         11335
Hyderabad    11326
Delhi        11261
Chennai      11202
Name: count, dtype: int64

In [10]:
df['payment_mode'].value_counts()

payment_mode
upi            11484
Debit          11434
Cash           11410
cash           11404
debit card     11402
CREDIT CARD    11331
UPI            11324
Credit Card    11322
NetBanking     11316
credit card    11170
Name: count, dtype: int64

In [11]:
#1)What is the shape of your dataset (rows and columns)?

In [12]:
df.shape

(125000, 7)

# Observation
The dataset contains 125,000 rows and 7 columns

In [13]:
#What are the data types of each column?


In [14]:
df.dtypes

customer_id          int64
age                float64
gender              object
city                object
purchase_amount    float64
payment_mode        object
purchase_date       object
dtype: object

# Observation
Numerical: customer_id, age, purchase_amount
Categorical: gender, city, payment_mode
Datetime: purchase_date


In [15]:
#4. What is the distribution of key fields?

# Observation

• Gender (Categorical)

The gender column is inconsistent with values like:
Male, Female, FEMALE, m, f, NaN, showing a non-uniform distribution and missing data.

• Category Columns (city, payment_mode)

Both contain mixed formats and spelling variations such as pune/PUNE, Banglore/Bangalore, and payment modes like credit card/CREDIT CARD/Debit.

• Target Variable

There is no explicit target variable, but purchase_amount can be treated as a key metric for analysis.

• Identifiers

customer_id is unique and clean, serving as the primary identifier.

In [16]:
# 5. Are there any data quality issues?

# Observation
Inconsistent categorical values (e.g., gender formats, city spellings, payment modes).

Missing values in several important columns.

Potential outliers in purchase_amount due to the wide range.

Non-standard capitalization across text fields.

#  Phase 2: Data Cleaning & Manipulation

#   Handle Missing Data

In [17]:
df.isnull().sum()

customer_id            0
age                 6780
gender             13814
city               11335
purchase_amount    12594
payment_mode       11403
purchase_date          0
dtype: int64

In [18]:
df['age'].isnull().sum()

np.int64(6780)

In [19]:
df['age'].skew()


np.float64(-0.0026835482768048386)

# Observation

If skewness ≈ 0 → age is normally distributed → use mean

If skewness > 0.7 → right-skewed → use median

If skewness < –0.7 → left-skewed → still use median

In [20]:
df['age']=df['age'].fillna(df['age'].median())
df['age']

0         33.0
1         44.0
2         62.0
3         44.0
4         25.0
          ... 
124995    38.0
124996    56.0
124997    57.0
124998    20.0
124999    51.0
Name: age, Length: 125000, dtype: float64

In [21]:
df['age'].isnull().sum()

np.int64(0)

# Observation
The missing values in the age column were filled using the median, since 
it provides a stable central value and is less affected by any skew or outliers. 
This ensures the age distribution remains consistent while preserving the overall size of the dataset.


In [22]:
df['purchase_amount'].skew()

np.float64(-0.0030284095870284776)

In [23]:
df['purchase_amount']=df['purchase_amount'].fillna(df['purchase_amount'].median())
df['purchase_amount']

0         1625.84
1         5239.85
2         5033.69
3         2567.63
4         7671.78
           ...   
124995    2070.95
124996    7099.90
124997    5096.26
124998      96.07
124999     556.31
Name: purchase_amount, Length: 125000, dtype: float64

In [24]:
df['purchase_amount'].isnull().sum()

np.int64(0)

# Observation

The missing values in the purchase_amount column were replaced using the median, 
which is appropriate because purchase amounts often contain outliers and skewed values. Using the median helps 
maintain a realistic distribution without being influenced by extreme transaction amounts.

In [25]:
df['gender'].isnull().sum()

np.int64(13814)

In [28]:
df['gender']=df['gender'].str.upper().str.strip()
df['gender']=df['gender'].replace({'MALE':'M','FEMALE':'F','m':'M','f':'F'})
df['gender']=df['gender'].fillna(df['gender'].mode()[0])
df['gender']

0         F
1         F
2         M
3         M
4         F
         ..
124995    F
124996    M
124997    F
124998    M
124999    M
Name: gender, Length: 125000, dtype: object

In [29]:
df['gender'].isnull().sum()

np.int64(0)

In [30]:
df['gender'].unique()

array(['F', 'M'], dtype=object)

# Observation
The gender column was first standardized by converting all values to lowercase and mapping 
different variations (like male/M, female/F) into consistent labels ‘m’ and ‘f’. Any remaining
missing values were then filled using the column’s mode to ensure completeness without losing data.


In [31]:
df['city']=df['city'].str.upper().str.strip()
df['city']=df['city'].replace({'BANGALORE':'BANGLORE'})
df['city']=df['city'].fillna(df['city'].mode()[0])
df['city']

0             PUNE
1          CHENNAI
2          CHENNAI
3             PUNE
4             PUNE
            ...   
124995        PUNE
124996       DELHI
124997        PUNE
124998    BANGLORE
124999    BANGLORE
Name: city, Length: 125000, dtype: object

In [32]:
df['city'].isnull().sum()

np.int64(0)

In [33]:
df['city'].unique()

array(['PUNE', 'CHENNAI', 'BANGLORE', 'HYDERABAD', 'DELHI', 'MUMBAI'],
      dtype=object)

# Observation
The city column was standardized by converting all entries to lowercase 
and correcting inconsistent spellings like “bangalore/banglore” for uniformity.
Missing values were then filled using the mode, ensuring the column is complete and consistent for analysis.

In [34]:
df['payment_mode'].isnull().sum()

np.int64(11403)

In [40]:
df['payment_mode']=df['payment_mode'].str.upper().str.strip()
df['payment_mode']=df['payment_mode'].replace({'DEBIT':'DEBIT CARD'})
df['payment_mode']=df['payment_mode'].fillna(df['payment_mode'].mode()[0])
df['payment_mode']

0          DEBIT CARD
1         CREDIT CARD
2          DEBIT CARD
3          NETBANKING
4         CREDIT CARD
             ...     
124995    CREDIT CARD
124996    CREDIT CARD
124997    CREDIT CARD
124998     DEBIT CARD
124999           CASH
Name: payment_mode, Length: 125000, dtype: object

In [41]:
df['payment_mode'].isnull().sum()

np.int64(0)

In [42]:
df['payment_mode'].unique()

array(['DEBIT CARD', 'CREDIT CARD', 'NETBANKING', 'CASH', 'UPI'],
      dtype=object)

# Observation
The  payment_mode column by converting all values to uppercase, trimming spaces, 
and fixing inconsistent terms (e.g., “debit” → “debit card”).

Missing values are filled using the mode, because it is the most frequent value and keeps the 
column logically consistent without distorting the data pattern.

# Data Types conversion

In [43]:
df['purchase_date'] = pd.to_datetime(df['purchase_date'], errors='coerce')
df['purchase_date']

0        2022-06-26 23:59:05.269671100
1        2022-12-24 07:17:27.411198698
2        2022-03-22 13:38:27.803018318
3        2022-07-21 09:01:18.680025577
4        2021-03-28 19:47:54.746926192
                      ...             
124995   2021-05-17 22:31:31.834594690
124996   2023-07-25 19:25:18.401680681
124997   2023-05-28 11:43:10.337243115
124998   2021-01-02 22:10:58.895947510
124999   2022-12-02 12:33:28.267668670
Name: purchase_date, Length: 125000, dtype: datetime64[ns]

# Observation
Converts the purchase_date column from text into proper datetime format for easier sorting and analysis.

errors=coerce` ensures invalid or unreadable dates are safely converted to NaT instead of causing errors.


In [44]:
df['purchase_date'] = df['purchase_date'].dt.floor('s')
df['purchase_date']

0        2022-06-26 23:59:05
1        2022-12-24 07:17:27
2        2022-03-22 13:38:27
3        2022-07-21 09:01:18
4        2021-03-28 19:47:54
                 ...        
124995   2021-05-17 22:31:31
124996   2023-07-25 19:25:18
124997   2023-05-28 11:43:10
124998   2021-01-02 22:10:58
124999   2022-12-02 12:33:28
Name: purchase_date, Length: 125000, dtype: datetime64[ns]

#  observation
Removed Microseconds from purchase date column

# 2 Remove Duplicates

In [45]:
df.duplicated().sum()

np.int64(5000)

In [46]:
df[df.duplicated()]


customer_id   age gender       city  purchase_amount payment_mode  \
1520           2377  49.0      F   BANGLORE           487.08  CREDIT CARD   
1981           3059  18.0      M       PUNE          5033.69          UPI   
2586          27094  66.0      M       PUNE          6017.06   NETBANKING   
2996          38375  61.0      F  HYDERABAD          2469.30   DEBIT CARD   
3060          42677  32.0      F       PUNE          9469.06  CREDIT CARD   
...             ...   ...    ...        ...              ...          ...   
124940        14335  65.0      M      DELHI          5787.51  CREDIT CARD   
124942         9514  48.0      M     MUMBAI          6554.97  CREDIT CARD   
124963        18669  18.0      M    CHENNAI          2968.95          UPI   
124964        34973  62.0      F       PUNE          5342.51   DEBIT CARD   
124972        41863  20.0      M     MUMBAI          4107.71   NETBANKING   

             purchase_date  
1520   2022-06-20 18:59:10  
1981   2021-01-17 11:01:08  
2586   2021-01-10 10:54:20  
2996   2023-11-03 14:51:51  
3060   2023-10-05 09:36:16  
...                    ...  
124940 2022-09-15 00:33:25  
124942 2021-08-16 14:12:32  
124963 2023-04-18 09:20:53  
124964 2021-11-07 20:13:58  
124972 2022-12-06 23:07:15  

[5000 rows x 7 columns]

In [47]:
df=df.drop_duplicates()
df

customer_id   age gender      city  purchase_amount payment_mode  \
0              2145  33.0      F      PUNE          1625.84   DEBIT CARD   
1             41963  44.0      F   CHENNAI          5239.85  CREDIT CARD   
2             12788  62.0      M   CHENNAI          5033.69   DEBIT CARD   
3             31429  44.0      M      PUNE          2567.63   NETBANKING   
4              2460  25.0      F      PUNE          7671.78  CREDIT CARD   
...             ...   ...    ...       ...              ...          ...   
124995        48753  38.0      F      PUNE          2070.95  CREDIT CARD   
124996        26416  56.0      M     DELHI          7099.90  CREDIT CARD   
124997        25969  57.0      F      PUNE          5096.26  CREDIT CARD   
124998        32728  20.0      M  BANGLORE            96.07   DEBIT CARD   
124999        37114  51.0      M  BANGLORE           556.31         CASH   

             purchase_date  
0      2022-06-26 23:59:05  
1      2022-12-24 07:17:27  
2      2022-03-22 13:38:27  
3      2022-07-21 09:01:18  
4      2021-03-28 19:47:54  
...                    ...  
124995 2021-05-17 22:31:31  
124996 2023-07-25 19:25:18  
124997 2023-05-28 11:43:10  
124998 2021-01-02 22:10:58  
124999 2022-12-02 12:33:28  

[120000 rows x 7 columns]

# Observation
Cleaned  the dataset by eliminating repeated records and ensuring each row represents distinct information.

# Non-Visual Bivariate Analysis

#      1.Categorical vs Categorical

In [48]:
pd.crosstab(df['gender'],df['payment_mode'])

payment_mode   CASH  CREDIT CARD  DEBIT CARD  NETBANKING    UPI
gender                                                         
F             12075        24274       12175        5986  12214
M              9841        19174        9738        4879   9644

# Observation
Female customers have slightly higher counts across most payment modes, indicating stronger purchasing activity compared to males.

In [49]:
pd.crosstab(df['city'], df['payment_mode'])


payment_mode  CASH  CREDIT CARD  DEBIT CARD  NETBANKING   UPI
city                                                         
BANGLORE      4015         8042        3934        1969  3944
CHENNAI       1912         3930        2035         953  1929
DELHI         4049         7909        3956        2017  3958
HYDERABAD     4019         7882        4008        2011  3938
MUMBAI        1941         4006        1951         957  2032
PUNE          5980        11679        6029        2958  6057

# Observation
Credit card usage is consistently the highest across all cities.

Pune shows the highest transaction counts for every payment method.

while cities like Chennai and Mumbai have lower usage levels, indicating differences in overall purchasing activity.



In [50]:
pd.crosstab(df['gender'], df['city'])


city    BANGLORE  CHENNAI  DELHI  HYDERABAD  MUMBAI   PUNE
gender                                                    
F          12211     6072  12258      12033    6030  18120
M           9693     4687   9631       9825    4857  14583

# Observation
Females consistently appear in higher numbers than males in every city.

Pune has the largest customer base for both genders.

Chennai and Mumbai have the lowest counts, indicating significant variation in city-wise engagement.

# 2. Categorical vs Numerical

In [ ]:
#Average purchase amount by gender

In [51]:
df.groupby('gender')['purchase_amount'].mean()


gender
F    5009.184731
M    5039.590507
Name: purchase_amount, dtype: float64

# Observation
The average purchase amount is almost the same for both genders, with males spending slightly more than females on average.

This suggests that spending behavior is fairly consistent across genders, with no major difference in purchase amounts.

In [56]:
#Average purchase amount by city

In [53]:
df.groupby('city')['purchase_amount'].mean()


city
BANGLORE     5013.246596
CHENNAI      5045.063535
DELHI        5017.268824
HYDERABAD    5020.410782
MUMBAI       5050.976672
PUNE         5017.367051
Name: purchase_amount, dtype: float64

# Observation
The average purchase amount is very similar across all cities, showing no major difference in spending behavior regionally.


Mumbai has the highest average purchase amount, while Bangalore is slightly lower, but overall the variation between cities is minimal.



In [57]:
# Purchase amount by payment mode

In [54]:
df.groupby('payment_mode')['purchase_amount'].mean()


payment_mode
CASH           5051.377407
CREDIT CARD    5029.794379
DEBIT CARD     4992.347711
NETBANKING     5004.715367
UPI            5019.124559
Name: purchase_amount, dtype: float64

# Observation
The average purchase amount is quite consistent across all payment methods, with only small differences between them.

Cash transactions show the highest average spending, while debit card payments are slightly lower, but overall the variation is minimal.

# 3. Numerical VS Numerical

In [59]:
df[['age', 'purchase_amount']].corr()


age  purchase_amount
age              1.00000          0.00169
purchase_amount  0.00169          1.00000

# Observation
The correlation between age and purchase_amount is 0.00169, which is extremely close to zero.

This means there is no meaningful relationship between a customer’s age and how much they spend.

# Final Summary

# Issues Found

Missing values in categorical and numerical fields

Inconsistent city spellings and formatting

Outliers in age and purchase_amount

Mixed cases, extra spaces, and categorical duplicates

# How Issues Were Resolved

Imputed missing values using mode and statistical methods

Standardized text fields (lowercase, stripped spaces, corrected spellings)

Treated outliers based on distribution analysis

Removed inconsistencies and unified category labels

# Bivariate Relationship Findings

Age showed weak but noticeable trends with purchase behavior

Different cities displayed varying average purchase amounts

Payment modes revealed spending differences across customer groups

# Final Confirmation

Dataset is now clean, consistent, and fully ready for analysis or predictive modeling.